vowpalwabbit/core/include/vw/core/reductions/automl/automl_impl.h

```diff
@@ -0,0 +1,318 @@
+        cm->pre_process(ec);
+        cm->config_oracle();
+        offset_learn(base, ec, logged, labelled_action);
+        current_state = automl_state::Experimenting;
+        break;
+
+      case automl_state::Experimenting:
+        cm->pre_process(ec);
+        cm->schedule();
+        offset_learn(base, ec, logged, labelled_action);
+        cm->update_champ();
+        break;
+
+      default:
+        break;
+    }
+  };
+  // inner loop of learn driven by # MAX_CONFIGS
+  void offset_learn(multi_learner& base, multi_ex& ec, CB::cb_class& logged, uint64_t labelled_action)
+  {
+    interaction_vec_t* incoming_interactions = ec[0]->interactions;
+    for (VW::example* ex : ec)
+    {
+      _UNUSED(ex);
+      assert(ex->interactions == incoming_interactions);
+    }
+
+    // if (ec[0]->interactions != nullptr)
+    // { logger->out_info("offlearn: incoming interaction: {}", ::interaction_vec_t_to_string(*(ec[0]->interactions)));
+    // }
+    const float w = logged.probability > 0 ? 1 / logged.probability : 0;
+    const float r = -logged.cost;
+
+    std::swap(ec[0]->pred.a_s, buffer_a_s);
+
+    int64_t current_slot_index = cm->scores.size() - 1;
+    int64_t live_slot = current_slot_index;
+    int64_t current_champ = static_cast<int64_t>(cm->current_champ);
+    assert(current_champ >= 0);
+
+    auto restore_guard = VW::scope_exit([this, &ec, &live_slot, &current_champ, &incoming_interactions]() {
+      for (example* ex : ec) { ex->interactions = incoming_interactions; }
+      if (live_slot >= 0 && live_slot != current_champ) { std::swap(ec[0]->pred.a_s, buffer_a_s); }
+    });
+
+    for (; current_slot_index >= 0; current_slot_index -= 1)
+    {
+      if (!debug_reverse_learning_order) { live_slot = current_slot_index; }
+      else
+      {
+        live_slot = cm->scores.size() - 1 - current_slot_index;
+      }
+
+      // TODO: what to do if that slot is switched with anew config?
+      std::swap(*_all_normalized, per_live_model_state_double[live_slot * 2]);
+      std::swap(*_gd_total_weight, per_live_model_state_double[live_slot * 2 + 1]);
+      std::swap(*_cb_adf_event_sum, per_live_model_state_uint64[live_slot * 2]);
+      std::swap(*_cb_adf_action_sum, per_live_model_state_uint64[live_slot * 2 + 1]);
+
+      if (live_slot == current_champ) { std::swap(ec[0]->pred.a_s, buffer_a_s); }
+      else
+      {
+        ec[0]->pred.a_s.clear();
+      }
+
+      for (example* ex : ec) { cm->apply_config(ex, live_slot); }
+
+      if (!base.learn_returns_prediction) { base.predict(ec, live_slot); }
+      base.learn(ec, live_slot);
+      const uint32_t chosen_action = ec[0]->pred.a_s[0].action;
+      if (live_slot == current_champ)
+      {
+        for (size_t live_slot_upd = 0; live_slot_upd < cm->scores.size(); ++live_slot_upd)
+        {
+          if (live_slot_upd == static_cast<size_t>(current_champ))
+          { cm->scores[live_slot_upd].first.update(chosen_action == labelled_action ? w : 0, r); }
```

Comment:
Approving for code review sync, may want to consider reverting since champ does not need statistics against itself

Reply:
corrected here https://github.com/VowpalWabbit/vowpal_wabbit/pull/4008